scripts/base/frameworks/storage/main.zeek

```diff
@@ -0,0 +1,170 @@
+##! The storage framework provides a way to store long-term data to disk.
+
+@load base/bif/storage.bif
+
+module Storage;
+
+export {
+	## Record for passing arguments to ``put``.
+	type PutArgs: record {
+		backend: opaque of Storage::BackendHandle;
+
+		# The script-level type of keys stored in the backend. Used for
+		# validation of keys passed to other framework methods.
+		key: any;
+
+		# The script-level type of keys stored in the backend. Used for
+		# validation of values passed to ``put`` as well for type
+		# conversions for return values from ``get``.
+		value: any;
+
+		# Indicates whether this value should overwrite an existing entry
+		# for the key.
+		overwrite: bool &default=F;
+
+		# An interval of time until the entry is automatically removed from the
+		# backend.
+		expire_time: interval &default=0sec;
+
+		# Indicates whether this operation should happen
+		# asynchronously. If this is true, the call to put must happen
+		# as part of a when statement. This flag is overridden when
+		# reading pcaps, since time won't move forward the same as when
+		# caputring live traffic.
+		async_mode: bool &default=T;
+	};
+
+	## Options record for the built-in SQLite backend.
+	type SqliteOptions: record {
+		## Path to the database file on disk
+		database_path: string;
+
+		## Name of the table used for storing data
+		table_name: string;
+
+		## Key/value table for passing tuning parameters when opening
+		## the database.  These must be pairs that can be passed to the
+		## ``pragma`` command in sqlite.
+		tuning_params: table[string] of string &default=table(
+			["journal_mode"] = "WAL",
+			["synchronous"] = "normal",
+			["temp_store"] = "memory"
+		);
+	};
+
+	## Options record for the built-in Redis backend.
+	type RedisOptions: record {
+		# Address to the server
+		server_addr: string;
+
+		# Port for the server
+		server_port: port;
+
+		# Server unix socket file. This can be used instead of the
+		# address and port above to connect to a local server.
+		server_unix_socket: string;
+
+		# Prefix used in key values stored to differentiate varying
+		# types of data on the same server.
+		key_prefix: string;
```

Comment:
Since values from different users can still collide I wonder whether this is useful at all, or should at least default to `""`

Reply:
There is a chance for collision in the key prefixes, but without it there'd also be a chance for collision of the keys themselves. I'd also argue that it should be user-configurable. It can't be something that's generated at runtime since if Zeek restarts it needs to know what the prefixes were during the last runtime.